Add doc on using NSX cluster networking (#111)
bcgov/platform-developer-docs: src/docs/platform-architecture-reference/nsx-networking.md

> # NSX SDN Cluster <!-- omit in toc -->

Reviewer: ```suggestion
---
title: NSX Networking 

slug: nsx-networking

description: 

keywords: NSX, SDN, cluster, SNAT,  openshift

page_purpose: 

audience: technical lead, developer

author: Steven Barre

content_owner: Steven Barre

sort_order: 
---

# NSX SDN Cluster <!-- omit in toc -->
```

## NetPol differences

### Egress

Egress is now managed by network policy in addition to Ingress. The deny by default policy now includes some rules to allow access to DNS, the internal image registry, and the Kube API.

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: platform-services-controlled-deny-by-default
  namespace: steve-test
spec:
  podSelector: {}
  egress:
    - ports:
        - protocol: TCP
          port: 5353
        - protocol: UDP
          port: 5353
      to:
        - podSelector:
            matchLabels:
              dns.operator.openshift.io/daemonset-dns: default
          namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: openshift-dns
    - ports:
      - port: 5000
        protocol: TCP
      to:
      - namespaceSelector:
          matchLabels:
            kubernetes.io/metadata.name: openshift-image-registry
        podSelector:
          matchLabels:
            docker-registry: default
    - ports:
      - port: 6443
        protocol: TCP
      to:

>       - ipBlock:
>           cidr: # IP of master server

Reviewer 2: ```suggestion
      - ipBlock:
          cidr: # IP of master server
    - ports:
      - port: 8080
        protocol: TCP
      to:
      - ipBlock:
          cidr: # IP of OCIO Forward Web Proxy, include full list         
```

Added forward proxy config for completeness.

Author: My understanding is we don't want access to the FP by default. I noted below in the FP section that teams need to add a netpol to allow it if they want internet access. This also lets them limit FP access to just a subset of pods and prevent a DB pod for example from being able to contact the internet,

Reviewer 2: fair enough. :)

Author: Added an example netpol for forward proxy

            matchLabels:
              app.kubernetes.io/name: vault-agent-injector
          namespaceSelector:
            matchLabels:
              kubernetes.io/metadata.name: openshift-bcgov-vault
    - ports:
        - protocol: TCP
          port: 443
      to:
        - ipBlock:
            cidr: 142.34.229.4/32
        - ipBlock:
            cidr: 142.34.64.4/32
  policyTypes:
    - Egress
```

## Forward Proxy

While the `*-tools` namespace has a public SNAT IP and can easily reach the internet, the rest of the namespaces cannot directly connect to the internet.

If you wish to connect to the internet, you must go via the Forward Proxy.

Kamloops: swpxkam.gov.bc.ca:8080 / 142.34.229.249:8080

Calgary: swpxcal.gov.bc.ca:8080 / 142.34.94.249:8080

You will also need a netpol to allow the egress from your pods to the forward proxy.

```yaml
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  name: django-egress-proxy
  namespace: steve-test
spec:
  egress:
  - to:
    - ipBlock:
        cidr: 142.34.229.249
    ports:
    - port: 8080
      protocol: TCP
  podSelector:
    matchLabels:
      name: django-psql-persistent
  policyTypes:
  - Egress
```

> Most applications recognize the env vars for proxies. Set them on your pod and see if things "just work".

Reviewer: ```suggestion
Most applications recognize the environment variables for proxies. Set them on your pod and see if things "just work".
```
